Source/ablastr/particles/DepositCharge.H

```diff
@@ -28,51 +30,72 @@ namespace particles {
+ * \param np_to_depose number of particles to depose (default: pti.numParticles())
+ * \param icomp component in MultiFab to start depositing to
+ * \param nc number of components to deposit
  * \param cost pointer to (load balancing) cost corresponding to box where present
-               particles deposit current. If nullptr, costs are not updated.
- * \param n_rz_azimuthal_modes number of azimuthal modes in use, irrelevant outside RZ geometry.
- * \param load_balance_costs_update_algo selected method for updating load balance costs.
- * \param do_device_synchronize call amrex::Gpu::synchronize() for tiny profiler regions
+               particles deposit current. If nullptr, costs are not updated. (default: nullptr)
+ * \param load_balance_costs_update_algo selected method for updating load balance costs (default: 0)
+ * \param do_device_synchronize call amrex::Gpu::synchronize() for tiny profiler regions (default: true)
  */
 template< typename T_PC >
 static void
 deposit_charge (typename T_PC::ParIterType& pti,
                 typename T_PC::RealVector const& wp,
-                const int * const ion_lev,
-                amrex::MultiFab* rho, const int icomp, const int nc,
-                const long offset, const long np_to_depose,
-                amrex::FArrayBox& local_rho, const int lev, const int depos_lev,
-                const amrex::Real charge, const int nox, const int noy, const int noz,
-                const amrex::IntVect& ng_rho, const std::array<amrex::Real,3>& dx,
-                const std::array<amrex::Real, 3>& xyzmin,
-                const amrex::IntVect& ref_ratio,
-                amrex::Real* cost, const int n_rz_azimuthal_modes,
-                const long load_balance_costs_update_algo,
-                const bool do_device_synchronize)
+                amrex::Real const charge,
+                int const * const ion_lev,
+                amrex::MultiFab* rho,
+                amrex::FArrayBox& local_rho,
+                int const particle_shape,
+                std::array<amrex::Real, 3> const & dx,
+                std::array<amrex::Real, 3> const & xyzmin,
+                int const n_rz_azimuthal_modes = 0,
+                std::optional<int> depos_lev = std::nullopt,
+                std::optional<amrex::IntVect> rel_ref_ratio = std::nullopt,
+                long const offset = 0,
+                std::optional<long> np_to_depose = std::nullopt,
+                int const icomp = 0, int const nc = 1,
+                amrex::Real * const AMREX_RESTRICT cost = nullptr,
+                long const load_balance_costs_update_algo = 0,
+                bool const do_device_synchronize = true)
 {
-    AMREX_ALWAYS_ASSERT_WITH_MESSAGE((depos_lev==(lev-1)) ||
-                                     (depos_lev==(lev  )),
-                                     "Deposition buffers only work for lev-1");
+    // deposition guards
+    amrex::IntVect const ng_rho = rho->nGrowVect();
```

Comment:
Is this truly identical to the previously passed `ng_rho` parameter?

Reply:
No, because here we allocate to the max...
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/WarpXParticleContainer.cpp#L753-L755
but then use something smaller to deposit:
https://github.com/ECP-WarpX/WarpX/blob/development/Source/Particles/WarpXParticleContainer.cpp#L597

@EZoni :
> No it's not the same because we explicitly want to deposit charge and currents in a number of guard cells that is less than all the guard cells that we allocate for the MultiFabs. So we set these different numbers of guard cells in the guardCellManager (the ones for deposition are called ng_depos_... and the ones for allocation and exchanges are called ng_alloc_...)
> When you call nGrowVect() on a MultiFab I think you get the ng_alloc_... effectively